crates/ruff_db/src/diagnostic/mod.rs

```diff
@@ -268,6 +301,167 @@ impl Diagnostic {
+    pub fn suggestion(&self) -> Option<&str> {
+        self.primary_annotation()?.get_message()
+    }
+
+    /// Returns the URL for the rule documentation, if it exists.
+    pub fn to_url(&self) -> Option<String> {
+        if self.is_syntax_error() {
+            None
+        } else {
+            Some(format!(
+                "{}/rules/{}",
+                env!("CARGO_PKG_HOMEPAGE"),
+                self.name()
+            ))
+        }
+    }
+
+    /// Returns the filename for the message.
+    ///
+    /// Panics if the diagnostic has no primary span, or if its file is not a `SourceFile`.
+    pub fn expect_ruff_filename(&self) -> String {
+        self.expect_primary_span()
+            .expect_ruff_file()
+            .name()
+            .to_string()
+    }
+
+    /// Computes the start source location for the message.
+    ///
+    /// Panics if the diagnostic has no primary span, if its file is not a `SourceFile`, or if the
+    /// span has no range.
+    pub fn expect_ruff_start_location(&self) -> LineColumn {
+        self.expect_primary_span()
+            .expect_ruff_file()
+            .to_source_code()
+            .line_column(self.expect_range().start())
+    }
+
+    /// Computes the end source location for the message.
+    ///
+    /// Panics if the diagnostic has no primary span, if its file is not a `SourceFile`, or if the
+    /// span has no range.
+    pub fn expect_ruff_end_location(&self) -> LineColumn {
+        self.expect_primary_span()
+            .expect_ruff_file()
+            .to_source_code()
+            .line_column(self.expect_range().end())
+    }
+
+    /// Returns the [`SourceFile`] which the message belongs to.
+    pub fn ruff_source_file(&self) -> Option<&SourceFile> {
+        self.primary_span_ref()?.as_ruff_file()
+    }
+
+    /// Returns the [`SourceFile`] which the message belongs to.
+    ///
+    /// Panics if the diagnostic has no primary span, or if its file is not a `SourceFile`.
+    pub fn expect_ruff_source_file(&self) -> SourceFile {
+        self.expect_primary_span().expect_ruff_file().clone()
+    }
+
+    /// Returns the [`TextRange`] for the diagnostic.
+    pub fn range(&self) -> Option<TextRange> {
+        self.primary_span()?.range()
+    }
+
+    /// Returns the [`TextRange`] for the diagnostic.
+    ///
+    /// Panics if the diagnostic has no primary span or if the span has no range.
+    pub fn expect_range(&self) -> TextRange {
+        self.range().expect("Expected a range for the primary span")
+    }
+}
+
+impl Ord for Diagnostic {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.partial_cmp(other).unwrap_or(std::cmp::Ordering::Equal)
+    }
+}
+
+impl PartialOrd for Diagnostic {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(
+            (self.ruff_source_file()?, self.range()?.start())
+                .cmp(&(other.ruff_source_file()?, other.range()?.start())),
+        )
+    }
 }
```

Comment:
I think this implementation is different from ty's ordering. I'd suggest to remove the `Ord` implementation and instead add a `start_ordering(&self, other: &Diagnostic)` method that returns an `Ordering` which you can use in places where you used `Ord` before.